Added module to manage manageiq tenants

evertmulder opened the pull request:
##### SUMMARY
Added a module to manage tenants in ManageIQ

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
manageiq_tenant

##### ANSIBLE VERSION
```
 ansible --version
ansible 2.6.1
  config file = None
  configured module search path = [u'/Users/emulder/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Mar 21 2018, 16:34:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
No module exists for creating tenants so I created on, based on manageiq_user

Commit:
Added module to manage manageiq tenants

Commit:
Fixed pep8 and module validation issues for manageiq_tenant

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_tenant.py:
```diff
@@ -0,0 +1,447 @@
+  manageiq_tenant:
+    name: 'Dep1'
+    description: 'Manufacturing department'
+    parent_id: 1
```
Is 1 assumed to be the `My Company` tenant?  how can one know it's 1, without prior assumptions and by only using this module?
=> Can you make the module take `parent_name` param and look it up, instead of id?

evertmulder replied:
The parent name is not unique. You can easily have the following tenant structure:

/root
/root/sub1
/root/sub1/sub1
/root/sub1/root
/root/root/sub1

An option would be in stead of specifying the tenant name let the user specify the full path something like: tenant_path: /root/sub1/sub2 (where the tenant name would be sub2)


And another situation. Suppose I want to change the name for an existing tenant. Perhaps adding an optional tenant_id that can be used to change a tenant name? What do you think?

cben replied:
I like the path idea.  `/root/sub1/sub2` is much easier than calling the module 3 times, with `register:` to get parent ids to follow the chain...
Is `/` (or another char) impossible in a name?  If not, could take a list of components (and maybe allow a string `'foo'` as shorthand for `['foo']` for convenience).

Great question about renaming!  I don't think we supported renaming in any other miq module.   @gtanzillo @abellotti @zgalor @yaacov @elad661 any ideas?
Perhaps add an `old_name` or `new_name` (or path) field.  Is there any existing pattern for this in Ansible?  Can it be idempotent?

evertmulder replied:
I know in the vmware_guest lets you supply the VM uuid to rename a VM, see https://docs.ansible.com/ansible/2.5/modules/vmware_guest_module.html, search for "Rename a VM (requires the VM's uuid)". So perhaps to support renaming the internal Id could be supplied.

If have tested if a special character can be used as a separator, but all chars slashes, backslashes are all valid for a tenant name. Having looked at the vmware_guest module they use the name_match attribute, in combination with the uuid. This could also be an approach:
1. First use tenant_id, if supplied (rename tenant if name differs)
2. Use parent_id, if supplied , and tenant name (if the tenant is not found under the supplied parent_id but was uniquely identified by the name move the tenant to the supplied parent_id)
3. Use tenant name, If multiple found use the name_match attribute to select the first found or last found

or use the tenant_path as a list. I don't think the shorthand if 1 name is supplied would work. It is unclear if you want to edit the root tenant or some other sub tenant I think...

evertmulder replied:
I ended up allowing the giving the option to supply the parent name with build-in uniqueness checks as an alternative to the parent_id. This should cover most situations I think.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_tenant.py:
```diff
@@ -0,0 +1,447 @@
+  manageiq_tenant:
+    state: 'absent'
+    name: 'Dep1'
+    parent_id: 1
```
I assume parent_id not necessary for deletion?

evertmulder replied:
The tenant name does not uniquely identifies a tenant, the name in combination of the parent does.

cben replied:
Ah :+1:.  I should have started by saying I don't know anything about tenants in manageiq :)

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_tenant.py:
```diff
@@ -0,0 +1,447 @@
+      url: 'http://127.0.0.1:3000'
+      username: 'admin'
+      password: 'smartvm'
+      verify_ssl: False
```
I think for `http:` connections `verify_ssl` is meaningless and can be omitted.

evertmulder replied:
True. Is would make sense to have an example with the option verify_ssl: False shown.

cben commented on lib/ansible/modules/remote_management/manageiq/manageiq_tenant.py:
```diff
@@ -0,0 +1,447 @@
+        """ Creates the tenant quotas in manageiq.
+
+        Returns:
+            result
```
duh :-)  Something like ``dict with `msg` and `changed`.`` would be more informative

evertmulder replied:
I will check if the comments are correct 👍

cben reviewed (commented):
Sorry, fell off the radar.
Some initial comments, not a full review yet.

cben commented:
This seems stuck for no good reason, I guess we're both waiting for each other :)
I'm on vacation this week but you can ping the other maintainers.

Basically I'm good with whatever direction you want to take this, though I feel some form of path is nicer than name + parent_id.

About renaming: your call whether you want to implement it now or leave for future.
Tenants don't have a uuid, but they do have both name & description, e.g.:
```
#<Tenant:0x00561439d461a8
 id: 1,
 ...
 name: "My Company",
 description: "Tenant for My Company",
```
Are both visible in the UI?  If only one is, then arguably there is little need to support renaming the invisible one, and we could use it as the stable id for renaming the other one.
If that doesn't work, then yeah I guess use the numeric id...

yaacov commented:
@evertmulder hi, current state looks good to me, is this ready for merge from your side ?

evertmulder commented:
@yaacov I need a bit more time to incorporate some of the things discussed with @cben. I have some (free) time issues, but I will get to it real soon now.

Commit:
Changes based on review
- Added parent attribute to specify the parent
- Cleanup return of tenant
- Update documentation on quotas

yaacov commented:
shipit

evertmulder commented:
@yaacov Does anyone have to do something to get the PR merged? Does it needs to have the shipit label?

> If the maintainer says shipit, the pull request is labeled shipit, whereupon the Core team assesses it for final merge

yaacov commented:
> @yaacov Does anyone have to do something to get the PR merged? Does it needs to have the shipit label?

> > If the maintainer says shipit, the pull request is labeled shipit, whereupon the Core team assesses it for final merge

@bcoca ^^  ?

bot_status

yaacov commented:
@evertmulder looks like we need more shipits ... @cben, @zgalor ?

cben commented:
shipit
Didn't really review the parent business, too complicated :)
I'll send some tiny doc fixes in follow-up PR.

Commit:
Updated version added to 2.8

yaacov commented:
shipit

:smile_cat:

cben commented:
shipit

gundalow commented:
Thank you for the PR and to all that reviewed. Merged into `devel` for release in Ansible 2.8.
